libass/ass_fontconfig.c

```diff
@@ -202,165 +184,78 @@ static void scan_fonts(FcConfig *config, ASS_FontProvider *provider)
-                    (FcChar8 **)&family);
-            if (result == FcResultMatch) {
-                return strdup(family);
-            } else {
-                return NULL;
-            }
-        }
+        FcLangSetDestroy(ls);
     }
 
-    // we shouldn't get here
-    return NULL;
-}
+    if (codepoint) {
+        FcCharSet *cs = FcCharSetCreate();
+        if (!cs)
+            goto cleanup;
 
-static void get_substitutions(void *priv, const char *name,
-                              ASS_FontProviderMetaData *meta)
-{
-    ProviderPrivate *fc = (ProviderPrivate *)priv;
+        FcCharSetAddChar(cs, codepoint);
+        FcPatternAddCharSet(pat, FC_CHARSET, cs);
 
-    FcPattern *pat = FcPatternCreate();
-    if (!pat)
-        return;
+        FcCharSetDestroy(cs);
+    }
 
-    FcPatternAddString(pat, FC_FAMILY, (const FcChar8 *)name);
-    FcPatternAddString(pat, FC_FAMILY, (const FcChar8 *)"__libass_delimiter");
+    FcPatternAddString(pat, FC_FAMILY, (const FcChar8 *)family);
     FcPatternAddBool(pat, FC_OUTLINE, FcTrue);
     if (!FcConfigSubstitute(fc->config, pat, FcMatchPattern))
         goto cleanup;
 
-    // read and strdup fullnames
-    meta->n_fullname = 0;
-    meta->fullnames = calloc(MAX_NAME, sizeof(char *));
-    if (!meta->fullnames)
-        goto cleanup;
+    FcDefaultSubstitute(pat);
+
+    if (!locale.len) {
+        // FC_LANG is automatically set according to locale, but this results
+        // in locale-specific fonts being used for text from other languages,
+        // which can look pretty bad (e.g. HannotateTC-W5 being used for English).
+        // We remove it here to use language-independent lookup instead if no language
+        // was specified by the file or the caller.
+        // Callers can pass in the system locale as a fallback behind a user setting
+        // if they want to.
+        FcPatternDel(pat, FC_LANG);
+    }
 
-    char *alias = NULL;
-    while (FcPatternGetString(pat, FC_FAMILY, meta->n_fullname,
-                (FcChar8 **)&alias) == FcResultMatch
-                && meta->n_fullname < MAX_NAME
-                && strcmp(alias, "__libass_delimiter") != 0) {
-        alias = strdup(alias);
-        if (!alias)
-            goto cleanup;
-        meta->fullnames[meta->n_fullname] = alias;
-        meta->n_fullname++;
+    FcPattern *match = FcFontMatch(fc->config, pat, &result);
+    if (match) {
+        FcChar8 *got_family = NULL;
+        result = FcPatternGetString(match, FC_FAMILY, 0, &got_family);
+        if (got_family)
+            ret = strdup((char *)got_family);
+
+        FcPatternDestroy(match);
     }
 
 cleanup:
-    FcPatternDestroy(pat);
+    if (pat)
+        FcPatternDestroy(pat);
+
+    return ret;
 }
 
 static ASS_FontProviderFuncs fontconfig_callbacks = {
     .check_postscript   = check_postscript,
     .check_glyph        = check_glyph,
     .destroy_provider   = destroy,
-    .get_substitutions  = get_substitutions,
```

Comment:
Ah, I've been testing this using `test/test.c`, which sets `default_family` to `"sans-serif"`. With that configuration and this patchset, the usual `serif`, `sans-serif`, and `monospace` aliases work fine for me. I get `DejaVuSerif`, `DejaVuSans`, and `DejaVuMono` for `en`, with CJK fallback on `MS-Mincho` or `MS-Gothic`; for `zh-CN`, I get `SimSun`, `ArialUnicodeMS`, and `NSimSun`; for `ja`, I get `MS-Mincho`, `MS-Gothic`, and `MS-Gothic`. This is all as expected.

With `Arial`, though, I get the same behavior you're describing here. I'm not sure what the best solution for this is; I can't see any way to implement the `get_substitutions` route in a locale-aware manner.

Reply:
Sorry to chime in here, was bothered by mpv's font selection process for a while, so I decided to do my investigation.
`FcFontMatch` is supposed to work with alias, because it performs partial match on `FC_FAMILY`, so `ass_fontconfig` won't fail at first instance for `{\fnserif}`.
For `Arial` I guess that's because @rcombs does not have `Arial Unicode MS` installed. That's the only `Arial` family font with CJK support out there.
IMO `get_substitutions` is not really useful with `fontconfig`, given that it already handles family alias well, but perhaps performing a search without locale if first search fails can handle this situation better.